[DOC] ShapeInference.md update. slyalin comments (#3355)
openvinotoolkit/openvino: docs/IE_DG/ShapeInference.md

==========================================

OpenVINO™ provides the following methods for runtime model reshaping:

* **Set a new input shape** with the `InferenceEngine::CNNNetwork::reshape` method.<br>
   The `InferenceEngine::CNNNetwork::reshape` method updates input shapes and propagates them down to the outputs of the model through all intermediate layers. 
   
> **NOTES**:
> - Starting with the 2021.1 release, the Model Optimizer converts topologies keeping shape-calculating sub-graphs by default, which enables correct shape propagation during reshaping in most cases.
> - Older versions of IRs are not guaranteed to reshape successfully. Please regenerate them with the Model Optimizer of the latest version of OpenVINO™.<br>
> - If an ONNX model does not have a fully defined input shape and the model was imported with the ONNX importer, reshape the model before loading it to the plugin.

* **Set a new batch dimension value** with the `InferenceEngine::CNNNetwork::setBatchSize` method.<br>     
   The meaning of a model batch may vary depending on the model design.
   This method doesn't deduce batch placement for inputs of the model from the model architecture.

>    It assumes batch is placed under the index 0 in the shape for all the inputs and uses `InferenceEngine::CNNNetwork::reshape` method to propagate updated shapes through the model.

Reviewer: ```suggestion
   It assumes that the batch is placed at the 0 index in the shape for all inputs and uses the `InferenceEngine::CNNNetwork::reshape` method to propagate updated shapes through the model.
```

   The method transforms the model before a new shape propagation to relax a hard-coded batch dimension in the model, if any.

   Use `InferenceEngine::CNNNetwork::reshape` instead of `InferenceEngine::CNNNetwork::setBatchSize` to set new input shapes for the model in case the model has:

>    * Multiple inputs with different zero index dimension meanings

Reviewer: ```suggestion
   * Multiple inputs with different zero-index dimension meanings
```

>    * Multiple inputs with different 0 index dimension meaning
>    * Input without a batch dimension
>    * scalar, 1D, 3D input or input without a batch dimension

Reviewer: ```suggestion
   * Multiple inputs with different 0 index dimension meanings
   * Input without a batch dimension
   * Scalar, 1D, or 3D input or input without a batch dimension
```

Doesn't the 3d point include the 2nd?

>    * Scalar, 1D, or 3D input or input without a batch dimension

Author: ```suggestion
   * 0D, 1D, or 3D shape
```

>    `InferenceEngine::CNNNetwork::setBatchSize` method is a high level API method which wraps `InferenceEngine::CNNNetwork::reshape` method call and works for trivial models from the batch placement standpoint.
>    Please use `InferenceEngine::CNNNetwork::reshape` for all the other models.

Reviewer: ```suggestion
   The `InferenceEngine::CNNNetwork::setBatchSize` method is a high-level API method that wraps the `InferenceEngine::CNNNetwork::reshape` method call and works for trivial models from the batch placement standpoint.
   Use `InferenceEngine::CNNNetwork::reshape` for other models.
```

>    Using `InferenceEngine::CNNNetwork::setBatchSize` method for models with a non-zero index batch placement or models with inputs without a batch dimension may lead to undefined behaviour.

Reviewer: ```suggestion
   Using the `InferenceEngine::CNNNetwork::setBatchSize` method for models with a non-zero index batch placement or for models with inputs that do not have a batch dimension may lead to undefined behaviour.
```
